front50-gcs/src/main/java/com/netflix/spinnaker/front50/model/InfrastructureDefinitionBucketDAO.java

```diff
@@ -0,0 +1,81 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.front50.model;
+
+import com.netflix.spinnaker.front50.exception.NotFoundException;
+import com.netflix.spinnaker.front50.model.infrastructureDefinition.InfrastructureDefinition;
+import com.netflix.spinnaker.front50.model.infrastructureDefinition.InfrastructureDefinitionDAO;
+import org.springframework.util.Assert;
+import rx.Scheduler;
+
+import java.util.Collection;
+import java.util.UUID;
+
+public class InfrastructureDefinitionBucketDAO extends BucketDAO<InfrastructureDefinition> implements InfrastructureDefinitionDAO {
+
+    public InfrastructureDefinitionBucketDAO(String basePath,
+                                             StorageService service,
+                                             Scheduler scheduler,
+                                             int refreshIntervalMs) {
+        super(InfrastructureDefinition.class, "infrastructureDefinitions",
+                basePath, service, scheduler, refreshIntervalMs);
+    }
+
+    @Override
+    public Collection<InfrastructureDefinition> getInfrastructureDefinitionHistory(String application, String account, int maxResults) {
+        String id = getInfrastructureDefinitionId(application, account);
+        return allVersionsOf(id, maxResults);
+    }
+
+    @Override
+    public String getInfrastructureDefinitionId(String application, String account) {
+        InfrastructureDefinition matched = all()
+                .stream()
+                .filter(infrastructureDefinition ->
+                        infrastructureDefinition.getAccount().equalsIgnoreCase(account) && infrastructureDefinition.getApplication().equalsIgnoreCase(application))
+                .findFirst()
+                .orElseThrow(() -> new NotFoundException(
+                        String.format("No infrastructure definition found with account '%s' in application '%s'.", account, application)
+                ));
+
+        return matched.getId();
+    }
+
+    @Override
+    public void update(String id, InfrastructureDefinition item) {
+        if (item.getId() == null) {
+            item.setId(id);
+        }
+        item.setTimestamp(String.valueOf(System.currentTimeMillis()));
+        super.update(id, item);
+    }
+
+    @Override
+    public InfrastructureDefinition create(String id, InfrastructureDefinition item) {
+        if (id == null) {
+            id = UUID.randomUUID().toString();
```

Comment:
Thoughts on not using a UUID for your ID ... it seems like you're really using `applicationName` + `account` as a composite identifier?

For example ... `Application` actually uses `applicationName` as it's id instead of a UUID.